1. Найдите посредством NumPy SVD для матрицы:

$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$


In [25]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)
A = np.array([[1, 2, 0], [0, 0, 5],[3, -4, 2],[1,6, 5], [0, 1, 0]])
print(f'Исходная матрица => \n{A}')

Исходная матрица => 
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [21]:
U, s, W = np.linalg.svd(A) # s - сингулярные числа
# Транспонируем матрицу W
V = W.T

# s - список диагональных элементов, его нужно привести к виду диагональной матрицы для наглядности
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

print(f'Матрица левых сингулярных векторов => \n {U}  ')
print(f'Матрица правых сингулярных векторов => \n {V}')
print(f'Диагональная матрица с сингулярными числами=> \n{D}')


Матрица левых сингулярных векторов => 
 [[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]  
Матрица правых сингулярных векторов => 
 [[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]
Диагональная матрица с сингулярными числами=> 
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]


2. Для матрицы из предыдущего задания найдите:

    а) евклидову норму;
    
    б) норму Фробениуса.
    
    
   **Решение**
   
   Евклидова норма определяет макстмальный коэффициент растяжения (сжатия). <br>
   В случае сингулярного разложения максимальное растяжение будет определяться максимальным сингулярным числом, <br>
   то есть $\left \| A \right \|_{E}= 8.82$ <br> <br>
   Норма Фробениуса равна корню квадратному от суммы квадратов сингулярных чисел : <br>
   $\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}} = \sqrt{8.82^2 +6.14^2+2.53^2} = 11.04$